src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt

```diff
@@ -0,0 +1,70 @@
+package nl.tudelft.ewi.se.ciselab.testgenie.actions
+
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.CommonDataKeys
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.editor.Caret
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.roots.ProjectRootManager
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiFile
+import nl.tudelft.ewi.se.ciselab.testgenie.evosuite.Runner
+
+/**
+ * This class contains all the logic related to generating tests for a line.
+ * No actual generation happens in this class, rather it is responsible for displaying the action option to the user when it is available,
+ *   getting the information about the selected class and passing it to (EvoSuite) Runner.
+ */
+class GenerateTestsActionLine : AnAction() {
+    private val log = Logger.getInstance(this.javaClass)
+
+    /**
+     * Performs test generation for a line when the action is invoked.
+     *
+     * @param e an action event that contains useful information
+     */
+    override fun actionPerformed(e: AnActionEvent) {
+        val project: Project = e.project ?: return
+
+        val psiFile: PsiFile = e.dataContext.getData(CommonDataKeys.PSI_FILE) ?: return
+        val caret: Caret = e.dataContext.getData(CommonDataKeys.CARET)?.caretModel?.primaryCaret ?: return
+
+        val psiClass: PsiClass = getSurroundingClass(psiFile, caret) ?: return
+        val classFQN = psiClass.qualifiedName ?: return
+
+        val selectedLine: Int = getSurroundingLine(psiFile, caret)?.plus(1) ?: return // lines in the editor and in EvoSuite are one-based
+        val vFile = e.dataContext.getData(CommonDataKeys.VIRTUAL_FILE) ?: return
+        val fileName = vFile.presentableUrl
+        val modificationStamp = vFile.modificationStamp
+
+        val projectPath: String = ProjectRootManager.getInstance(project).contentRoots.first().path
+        val projectClassPath = "$projectPath/target/classes/"
+
+        log.info("Generating tests for project $projectPath with classpath $projectClassPath")
+
+        log.info("Selected class is $classFQN")
+        log.info("Selected line is $selectedLine")
+
+        Runner(project, projectPath, projectClassPath, classFQN, fileName, modificationStamp).forLine(selectedLine)
+            .runEvoSuite()
+    }
```

Comment:
Since much of the code in this function is duplicated in all 3 test generation actions (class, method and line) is it possible to abstract it in some way, either by making a parent class or by moving it into a helper?

Reply:
No

Reply:
It would be useless since otherwise we will have to create methods with huge parameter lists or many return types. This will also not be good

Reply:
At least i don't see a decent solution for it

Reply:
Because all the duplicated lines are just getting the stuff from the event

Reply:
Great, thanks!